LibWeb: Begin implementing media resource seeking

trflynn89 opened the pull request:
https://user-images.githubusercontent.com/5600524/231310908-ede52697-6ea4-404a-99d1-fdf9f274a420.mp4

<br />This also gives us looping:


https://user-images.githubusercontent.com/5600524/231310927-77644388-e02c-4ce0-8840-033c6e6245a9.mp4

trflynn89 commented on Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:
```diff
@@ -1033,6 +1039,79 @@ WebIDL::ExceptionOr<void> HTMLMediaElement::pause_element()
+        return;
+
+    // FIXME: 3. If the element's seeking IDL attribute is true, then another instance of this algorithm is already running.
+    //           Abort that other instance of the algorithm without waiting for the step that it is running to complete.
```
This spec step is more about how we are supposed to be executing web spec steps in parallel, with the ability to abort them at any time. That whole scenario is one giant FIXME, as we execute everything synchronously right now.

Zaggy1024 replied:
Ah, hmm, I hadn't even considered in my implementation that `PlaybackManager` could have clients calling functions from multiple threads. In that case, perhaps it is good that the logic stay here to keep things simple in the manager.

trflynn89 commented on Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:
```diff
@@ -207,10 +206,12 @@ void HTMLMediaElement::set_duration(double duration)
         });
+
+        if (m_current_playback_position > duration)
+            seek_element(duration);
```
As of right now this is actually unreachable anyways. We only enter `set_duration` with a valid duration just after creating a `PlaybackManager`, and we set `m_current_playback_position` to 0 just before.

There's probably other spec steps we haven't implemented yet that would cause this to be reached, so when we get there, we'll see how it behaves I guess

Zaggy1024 replied:
I figured that was probably the case, no big deal for now. :^)

trflynn89 commented on Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:
```diff
@@ -1033,6 +1039,79 @@ WebIDL::ExceptionOr<void> HTMLMediaElement::pause_element()
+    }
+
+    // 4. Set the seeking IDL attribute to true.
+    m_seeking = true;
```
We can't assume we have a video element here, since this `HTMLMediaElement` (in the future) could also be a `HTMLAudioElement`. So tracking a boolean for the attribute seems fine.

Zaggy1024 replied:
In the future, the goal is for PlaybackManager to handle both video and audio (in any combination of one or both), so it should always be there to query. If it's used for tracking multiple seek requests across threads, though, that's fine with me.

trflynn89 commented on Userland/Libraries/LibWeb/HTML/HTMLVideoElement.cpp:
```diff
@@ -101,4 +101,10 @@ void HTMLVideoElement::on_paused()
+void HTMLVideoElement::on_seek(double position, MediaSeekMode seek_mode)
+{
+    if (m_video_track)
+        m_video_track->seek(Time::from_milliseconds(position * 1000.0), seek_mode);
```
Yeah I'll play around with it more after your PR lands, I had a waiting thing sketched out originally, so if it's best to keep that I'll add it back

trflynn89 commented on Userland/Libraries/LibVideo/PlaybackManager.cpp:
```diff
@@ -628,6 +628,7 @@ class PlaybackManager::SeekingStateHandler : public PlaybackManager::ResumingSta
                 }
 
                 manager().m_next_frame.emplace(item);
+                manager().m_last_present_in_media_time = m_target_timestamp;
```
Indeed it is!

trflynn89 commented:
Updates:
* Removed LibVideo fix for looping, as that was resolved in #18300
* Removed `Video::PlaybackManager::on_end_of_stream` in favor of the improved `Video::PlaybackManager::on_playback_state_change`
* More correctly implemented the seek step which says to wait until the underlying stream has finished seeking internally

Zaggy1024 commented:
Sorry, thought a maintainer would want to review it now. This all looks good to me, it works quite nicely locally for me, so feel free to merge!
